File: Task/Kolakoski-sequence/Java/kolakoski-sequence.java
import java.util.Arrays;

public class Kolakoski {
    private static class Crutch {
        final int len;
        int[] s;
        int i;

        Crutch(int len) {
            this.len = len;
            s = new int[len];
            i = 0;
        }

        void repeat(int count) {
            for (int j = 0; j < count; j++) {
                if (++i == len) return;
                s[i] = s[i - 1];
            }
        }
    }

    private static int nextInCycle(final int[] self, int index) {
        return self[index % self.length];
    }

    private static int[] kolakoski(final int[] self, int len) {
        Crutch c = new Crutch(len);

        int k = 0;
        while (c.i < len) {
            c.s[c.i] = nextInCycle(self, k);
            if (c.s[k] > 1) {
                c.repeat(c.s[k] - 1);
            }
            if (++c.i == len) return c.s;
            k++;
        }
        return c.s;
    }

    private static boolean possibleKolakoski(final int[] self) {
        int[] rle = new int[self.length];
        int prev = self[0];
        int count = 1;
        int pos = 0;
        for (int i = 1; i < self.length; i++) {
            if (self[i] == prev) {
                count++;
            } else {
                rle[pos++] = count;
                count = 1;
                prev = self[i];
            }
        }
        // no point adding final 'count' to rle as we're not going to compare it anyway
        for (int i = 0; i < pos; i++) {
            if (rle[i] != self[i]) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        int[][] ias = new int[][]{
            new int[]{1, 2},
            new int[]{2, 1},
            new int[]{1, 3, 1, 2},
            new int[]{1, 3, 2, 1}
        };
        int[] lens = new int[]{20, 20, 30, 30};

        for (int i=0; i<ias.length; i++) {
            int len = lens[i];
            int[] kol = kolakoski(ias[i], len);

            System.out.printf("First %d members of the sequence generated by %s: \n", len, Arrays.toString(ias[i]));
            System.out.printf("%s\n", Arrays.toString(kol));
            System.out.printf("Possible Kolakoski sequence? %s\n\n", possibleKolakoski(kol));
        }
    }
}

File: Task/Kolakoski-sequence/PascalABC.NET/kolakoski-sequence.pas
function nextInCycle(a: array of integer; index: word) :=  a[index mod a.Length];

function kolakoski(a: array of integer; length: word): array of integer;
begin
  SetLength(result, length);
  var i := 0;
  var k := 0;

  while true do
  begin
    result[i] := nextInCycle(a, k);
    if result[k] > 1 then
      for var j := 1 to result[k] - 1 do
      begin
        i += 1;
        if i = length then exit;
        result[i] := result[i - 1];
      end;
    i += 1;
    if i = length then exit;
    k += 1;
  end;
end;

function possibleKolakoski(a: array of integer): boolean;
begin
  var rle := new List<integer>;
  var prev := a[0];
  var count := 1;
  result := true;

  for var i := 1 to a.Count - 1 do
    if a[i] = prev then
      count += 1
    else
      begin
      rle.Add(count);
      count := 1;
      prev := a[i];
    end;

  foreach var val in rle index i do
    if val <> a[i] then result := false;
end;


begin
  var Ias := |[1, 2], [2, 1], [1, 3, 1, 2], [1, 3, 2, 1]|;
  var Lengths := [20, 20, 30, 30];

  foreach var (length, ia) in Zip(Lengths, Ias) do
  begin
    var kol := kolakoski(ia, length);
    writeln('First ', length, ' members of the sequence generated by ', ia, ':');
    writeln(kol);
    var s := if possibleKolakoski(kol) then 'Yes' else 'No';
    writeln('Possible Kolakoski sequence? ', s, #10);
  end;
end.
